File: crates/valka-sdk/src/context.rs
use std::collections::VecDeque;

use tokio::sync::mpsc;
use valka_proto::{LogEntry, SignalAck, TaskSignal, WorkerRequest, worker_request};

/// Data from a received signal.
pub struct SignalData {
    pub signal_id: String,
    pub name: String,
    pub payload: String,
}

impl SignalData {
    /// Parse the signal payload JSON into a typed value.
    pub fn parse_payload<T: serde::de::DeserializeOwned>(&self) -> Result<T, serde_json::Error> {
        serde_json::from_str(&self.payload)
    }
}

/// Context passed to task handlers, providing logging, metadata, and signal access.
pub struct TaskContext {
    pub task_id: String,
    pub task_run_id: String,
    pub queue_name: String,
    pub task_name: String,
    pub attempt_number: i32,
    pub input: String,
    pub metadata: String,
    request_tx: mpsc::Sender<WorkerRequest>,
    signal_rx: mpsc::Receiver<TaskSignal>,
    signal_buffer: VecDeque<TaskSignal>,
}

impl TaskContext {
    pub fn new(
        task_id: String,
        task_run_id: String,
        queue_name: String,
        task_name: String,
        attempt_number: i32,
        input: String,
        metadata: String,
        request_tx: mpsc::Sender<WorkerRequest>,
        signal_rx: mpsc::Receiver<TaskSignal>,
    ) -> Self {
        Self {
            task_id,
            task_run_id,
            queue_name,
            task_name,
            attempt_number,
            input,
            metadata,
            request_tx,
            signal_rx,
            signal_buffer: VecDeque::new(),
        }
    }

    /// Parse the input JSON
    pub fn input<T: serde::de::DeserializeOwned>(&self) -> Result<T, serde_json::Error> {
        serde_json::from_str(&self.input)
    }

    /// Wait for a signal with a specific name. Non-matching signals are buffered.
    pub async fn wait_for_signal(&mut self, name: &str) -> Option<SignalData> {
        // Check buffer first
        if let Some(idx) = self
            .signal_buffer
            .iter()
            .position(|s| s.signal_name == name)
        {
            let signal = self.signal_buffer.remove(idx).unwrap();
            self.send_signal_ack(&signal.signal_id).await;
            return Some(SignalData {
                signal_id: signal.signal_id,
                name: signal.signal_name,
                payload: signal.payload,
            });
        }

        // Wait for matching signal from channel
        while let Some(signal) = self.signal_rx.recv().await {
            if signal.signal_name == name {
                self.send_signal_ack(&signal.signal_id).await;
                return Some(SignalData {
                    signal_id: signal.signal_id,
                    name: signal.signal_name,
                    payload: signal.payload,
                });
            }
            // Buffer non-matching signals
            self.signal_buffer.push_back(signal);
        }

        None
    }

    /// Receive the next signal (any name). Checks buffer first.
    pub async fn receive_signal(&mut self) -> Option<SignalData> {
        // Check buffer first
        if let Some(signal) = self.signal_buffer.pop_front() {
            self.send_signal_ack(&signal.signal_id).await;
            return Some(SignalData {
                signal_id: signal.signal_id,
                name: signal.signal_name,
                payload: signal.payload,
            });
        }

        // Wait for next signal from channel
        if let Some(signal) = self.signal_rx.recv().await {
            self.send_signal_ack(&signal.signal_id).await;
            return Some(SignalData {
                signal_id: signal.signal_id,
                name: signal.signal_name,
                payload: signal.payload,
            });
        }

        None
    }

    async fn send_signal_ack(&self, signal_id: &str) {
        let request = WorkerRequest {
            request: Some(worker_request::Request::SignalAck(SignalAck {
                signal_id: signal_id.to_string(),
            })),
        };
        let _ = self.request_tx.send(request).await;
    }

    /// Log a message at INFO level
    pub async fn log(&self, message: &str) {
        self.log_at_level(2, message).await;
    }

    /// Log a message at DEBUG level
    pub async fn debug(&self, message: &str) {
        self.log_at_level(1, message).await;
    }

    /// Log a message at WARN level
    pub async fn warn(&self, message: &str) {
        self.log_at_level(3, message).await;
    }

    /// Log a message at ERROR level
    pub async fn error(&self, message: &str) {
        self.log_at_level(4, message).await;
    }

    async fn log_at_level(&self, level: i32, message: &str) {
        let entry = LogEntry {
            task_run_id: self.task_run_id.clone(),
            timestamp_ms: chrono::Utc::now().timestamp_millis(),
            level,
            message: message.to_string(),
            metadata: String::new(),
        };

        let batch = valka_proto::LogBatch {
            entries: vec![entry],
        };

        let request = WorkerRequest {
            request: Some(worker_request::Request::LogBatch(batch)),
        };

        let _ = self.request_tx.send(request).await;
    }
}

File: sdks/python/src/valka/context.py
"""Task execution context passed to handler functions."""

from __future__ import annotations

import asyncio
import json
import time
from dataclasses import dataclass
from typing import Any, Callable, Awaitable, TYPE_CHECKING

if TYPE_CHECKING:
    from valka._proto.valka.v1 import worker_pb2


@dataclass
class SignalData:
    """Data from a received signal."""

    signal_id: str
    name: str
    payload: str

    def parse_payload(self) -> Any:
        """Parse the signal payload JSON. Returns None if empty."""
        if not self.payload:
            return None
        return json.loads(self.payload)


class TaskContext:
    """Context provided to task handler functions.

    Exposes task metadata, logging, and signal reception methods.
    """

    def __init__(
        self,
        *,
        task_id: str,
        task_run_id: str,
        queue_name: str,
        task_name: str,
        attempt_number: int,
        raw_input: str,
        raw_metadata: str,
        send_fn: Callable[[worker_pb2.WorkerRequest], Awaitable[None]],
    ) -> None:
        self.task_id = task_id
        self.task_run_id = task_run_id
        self.queue_name = queue_name
        self.task_name = task_name
        self.attempt_number = attempt_number
        self._raw_input = raw_input
        self._raw_metadata = raw_metadata
        self._send_fn = send_fn
        self._signal_queue: asyncio.Queue[Any] = asyncio.Queue()
        self._signal_buffer: list[Any] = []

    def input(self) -> Any:
        """Parse and return the task input JSON. Returns None if empty."""
        if not self._raw_input:
            return None
        return json.loads(self._raw_input)

    def metadata(self) -> dict[str, Any]:
        """Parse and return the task metadata JSON. Returns {} if empty."""
        if not self._raw_metadata:
            return {}
        return json.loads(self._raw_metadata)

    async def log(self, message: str) -> None:
        """Send an INFO log entry."""
        await self._log_at_level(2, message)

    async def debug(self, message: str) -> None:
        """Send a DEBUG log entry."""
        await self._log_at_level(1, message)

    async def warn(self, message: str) -> None:
        """Send a WARN log entry."""
        await self._log_at_level(3, message)

    async def error(self, message: str) -> None:
        """Send an ERROR log entry."""
        await self._log_at_level(4, message)

    async def wait_for_signal(self, name: str) -> SignalData:
        """Wait for a signal with the given name. Non-matching signals are buffered."""
        # Check buffer first
        for i, sig in enumerate(self._signal_buffer):
            if sig.signal_name == name:
                self._signal_buffer.pop(i)
                await self._send_signal_ack(sig.signal_id)
                return SignalData(
                    signal_id=sig.signal_id,
                    name=sig.signal_name,
                    payload=sig.payload,
                )

        # Wait for matching signal
        while True:
            sig = await self._signal_queue.get()
            if sig.signal_name == name:
                await self._send_signal_ack(sig.signal_id)
                return SignalData(
                    signal_id=sig.signal_id,
                    name=sig.signal_name,
                    payload=sig.payload,
                )
            self._signal_buffer.append(sig)

    async def receive_signal(self) -> SignalData:
        """Wait for the next signal (any name). Checks buffer first."""
        if self._signal_buffer:
            sig = self._signal_buffer.pop(0)
            await self._send_signal_ack(sig.signal_id)
            return SignalData(
                signal_id=sig.signal_id,
                name=sig.signal_name,
                payload=sig.payload,
            )

        sig = await self._signal_queue.get()
        await self._send_signal_ack(sig.signal_id)
        return SignalData(
            signal_id=sig.signal_id,
            name=sig.signal_name,
            payload=sig.payload,
        )

    def _deliver_signal(self, signal: Any) -> None:
        """Internal: called by the worker to deliver a signal to this context."""
        self._signal_queue.put_nowait(signal)

    async def _send_signal_ack(self, signal_id: str) -> None:
        from valka._proto.valka.v1 import worker_pb2

        request = worker_pb2.WorkerRequest(
            signal_ack=worker_pb2.SignalAck(signal_id=signal_id)
        )
        await self._send_fn(request)

    async def _log_at_level(self, level: int, message: str) -> None:
        from valka._proto.valka.v1 import worker_pb2

        entry = worker_pb2.LogEntry(
            task_run_id=self.task_run_id,
            timestamp_ms=int(time.time() * 1000),
            level=level,
            message=message,
        )
        request = worker_pb2.WorkerRequest(
            log_batch=worker_pb2.LogBatch(entries=[entry])
        )
        await self._send_fn(request)
